pwiz_tools/Shared/Common/SystemUtil/StringSearch.cs

```diff
@@ -0,0 +1,463 @@
+            /// <param name="c">Character</param>
+            /// <returns>True if transition exists</returns>
+            public bool ContainsTransition(char c)
+            {
+                return GetTransition(c)!=null;
+            }
+
+            #endregion
+            #region Properties
+            
+            private char _char;
+            private TreeNode _parent;
+            private TreeNode _failure;
+            private ArrayList _results;
+            private TreeNode[] _transitionsAr;
+            private string[] _resultsAr;
+            private Hashtable _transHash;
+
+            /// <summary>
+            /// Character
+            /// </summary>
+            public char Char
+            {
+                get { return _char; }
+            }
+
+
+            /// <summary>
+            /// Parent tree node
+            /// </summary>
+            public TreeNode Parent
+            {
+                get { return _parent; }
+            }
+
+
+            /// <summary>
+            /// Failure function - descendant node
+            /// </summary>
+            public TreeNode Failure
+            {
+                get { return _failure; }
+                set { _failure=value; } 
+            }
+
+
+            /// <summary>
+            /// Transition function - list of descendant nodes
+            /// </summary>
+            public TreeNode[] Transitions
+            {
+                get { return _transitionsAr; }
+            }
+
+
+            /// <summary>
+            /// Returns list of patterns ending by this letter
+            /// </summary>
+            public string[] Results
+            {
+                get { return _resultsAr; }
+            }
+
+            #endregion
+        }
+
+        #endregion
+        #region Local fields
+        
+        /// <summary>
+        /// Root of keyword tree
+        /// </summary>
+        private TreeNode _root;
+
+        /// <summary>
+        /// Keywords to search for
+        /// </summary>
+        private IEnumerable<string> _keywords;
+
+        #endregion
+
+        #region Initialization
+
+        /// <summary>
+        /// Initialize search algorithm (Build keyword tree)
+        /// </summary>
+        /// <param name="keywords">Keywords to search for</param>
+        /// <param name="cancelToken">A token that can be used to abort building the Aho-Corasick tree</param>
+        public StringSearch(IEnumerable<string> keywords, CancellationToken? cancelToken = null)
+        {
+            SetKeywords(keywords, cancelToken);
+        }
+
+
+        /// <summary>
+        /// Initialize search algorithm with no keywords
+        /// (Use Keywords property)
+        /// </summary>
+        public StringSearch()
+        { }
+
+        #endregion
+        #region Implementation
+
+        /// <summary>
+        /// Build tree from specified keywords
+        /// </summary>
+        void BuildTree(CancellationToken? cancelToken)

```

Comment:
This is a private function. You mean to make this change to:
`public StringSearch(IEnumerable<string> keywords, CancellationToken? cancelToken = null)` ?
I could do that with an additional constructor rather than a default argument.